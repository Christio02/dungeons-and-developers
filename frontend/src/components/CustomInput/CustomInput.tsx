import React, { useEffect, useRef, useState } from 'react';
import SaveIcon from '@mui/icons-material/Save';
import ModeEditOutlineIcon from '@mui/icons-material/ModeEditOutline';

type InputProps = {
  placeholder: string;
  inputName: string;
  value: string;
  onSave: (newValue: string) => void;
};
/**
 * CustomInput component allows editing and saving a text value in place with a toggle between view and edit modes.
 * It displays a value, and when clicked, switches to an input field for editing.
 *
 * @component
 * @param {Object} props - The properties passed to the component.
 * @param {string} props.placeholder - Placeholder text shown when the input field is empty.
 * @param {string} props.inputName - The label/name for the input field (used for accessibility and alerts).
 * @param {string} props.value - The initial value to display.
 * @param {Function} props.onSave - Callback function triggered when the input value is saved.
 *                                   It receives the new value as an argument.
 *
 * @example
 * // Usage example:
 * <CustomInput
 *   placeholder="Enter your name"
 *   inputName="Name"
 *   value="John Doe"
 *   onSave={(newValue) => console.log(newValue)}
 * />
 */

const CustomInput = ({ placeholder, inputName, value, onSave }: InputProps) => {
  const [inputValue, setInputValue] = useState(value);

  const [isEditing, setIsEditing] = useState(false);

  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    setInputValue(value);
  }, [value]);

  useEffect(() => {
    if (isEditing && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isEditing]);

  /**
   * Handles logic when entering editMode
   */
  const handleEditClick = () => {
    setIsEditing(true);
  };

  /**
   * When user clicks saveIcon while in editing-mode, this verifies that value is not empty or equal to placeholder
   * If true, then alerts user, if not then triggers onSave callback function
   */

  const handleSaveClick = () => {
    const currentValue = inputValue.trim();

    if (currentValue === '' || currentValue === placeholder) {
      alert(`${inputName} cannot be empty!`);
      setInputValue(placeholder);
    } else {
      onSave(currentValue);
    }
    setIsEditing(false);
  };

  /**
   * Saves value when input loses focus, i.e. user clicks outside input field (onBlur event)
   */
  const handleBlur = () => {
    handleSaveClick();
  };

  /**
   * Handles the change event when user are writing inside input, limits value to 20 characters
   * @param{React.ChangeEvent<HTMLInputElement>} e - The change event
   */

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    let newName = e.target.value;

    if (newName.length > 20) {
      newName = newName.substring(0, 20);
    }
    setInputValue(newName);
  };

  return (
    <>
      <section className="flex flex-row items-center justify-center gap-4 ">
        {isEditing ? (
          <>
            <label htmlFor={inputName} className="sr-only">
              {inputName}
            </label>
            <input
              ref={inputRef}
              id={inputName}
              type="text"
              value={inputValue}
              aria-label="Edit name"
              data-testid="dungeon-input"
              onChange={handleChange}
              onBlur={handleBlur}
              autoFocus
              className="border-none rounded-lg outline-none focus:outline-none focus:ring-2 focus:ring-red-500 bg-transparent w-[65vw] lg:w-fit lg:p-y-5 text-6xl md:text-5xl lg:text-4xl xl:text-3xl 2xl:text-2xl text-center"
            />
          </>
        ) : (
          <h2
            className=" text-6xl md:text-5xl lg:text-4xl xl:text-3xl 2xl:text-2xl"
            aria-label="Name of your dungeon"
            data-testid="dungeon-name"
          >
            {inputValue}
          </h2>
        )}

        <button
          className=" p-[2px] lg:p-[5px] rounded-lg bg-customRed hover:bg-transparent border-2 border-customRed hover:border-customRed transition-colors duration-100 group"
          aria-label={isEditing ? `Save ${inputName}` : `Edit ${inputName}`}
          onClick={isEditing ? handleSaveClick : handleEditClick}
        >
          {isEditing ? (
            <SaveIcon className="text-white group-hover:text-customRed duration-100 lg:!w-6 lg:!h-6 xl:!w-8 xl:!h-8" />
          ) : (
            <ModeEditOutlineIcon className="text-white group-hover:text-customRed duration-100 lg:!w-6 lg:!h-6 xl:!w-8 xl:!h-8" />
          )}
        </button>
      </section>
    </>
  );
};

export default CustomInput;
