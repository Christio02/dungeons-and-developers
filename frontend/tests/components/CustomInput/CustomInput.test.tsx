import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import userEvent from '@testing-library/user-event';
import CustomInput from '../../../src/components/CustomInput/CustomInput';

describe('CustomInput Component', () => {
  const placeholder = 'Enter your dungeon name';
  const inputName = 'Dungeon';
  const initialValue = 'My dungeon';
  const onSaveMock = vi.fn();

  it('renders correctly in view mode', () => {
    render(<CustomInput placeholder={placeholder} inputName={inputName} value={initialValue} onSave={onSaveMock} />);
    expect(screen.getByTestId('dungeon-name')).toBeTruthy();
    expect(screen.getByRole('button', { name: `Edit ${inputName}` })).toBeTruthy();
  });

  it('renders correctly in edit mode', async () => {
    const user = userEvent.setup();
    render(<CustomInput placeholder={placeholder} inputName={inputName} value={initialValue} onSave={onSaveMock} />);

    const editButton = screen.getByRole('button', { name: `Edit ${inputName}` });
    await user.click(editButton);

    const input = screen.getByRole('textbox', { name: 'Edit name' });
    expect(input).toBeTruthy();
    const saveButton = screen.getByRole('button', { name: `Save ${inputName}` });
    expect(saveButton).toBeTruthy();
  });

  it('updates input value on change', async () => {
    const user = userEvent.setup();
    render(<CustomInput placeholder={placeholder} inputName={inputName} value={initialValue} onSave={onSaveMock} />);

    await user.click(screen.getByRole('button', { name: `Edit ${inputName}` }));
    const inputElement = screen.getByRole('textbox', { name: 'Edit name' }) as HTMLInputElement;
    await user.type(inputElement, 'Test');
    expect(inputElement.value).toBe(`${initialValue}Test`);
  });

  it('limits input value to 20 characters', async () => {
    const user = userEvent.setup();
    render(<CustomInput placeholder={placeholder} inputName={inputName} value={initialValue} onSave={onSaveMock} />);

    await user.click(screen.getByRole('button', { name: `Edit ${inputName}` }));
    const inputElement = screen.getByRole('textbox', { name: 'Edit name' });

    const longValue = 'aaaaaaaaaaaaaaaaaaaaaaaa';
    const expectedTrimmedValue = longValue.substring(0, 20);

    await user.clear(inputElement);

    await user.type(inputElement, longValue);

    const actualValue = (inputElement as HTMLInputElement).value;
    console.log(actualValue, actualValue.length);
    expect(actualValue).toHaveLength(20);

    expect(actualValue).toBe(expectedTrimmedValue);
  });

  it('calls onSave with the correct value when saved', async () => {
    const user = userEvent.setup();
    render(<CustomInput placeholder={placeholder} inputName={inputName} value={initialValue} onSave={onSaveMock} />);

    await user.click(screen.getByRole('button', { name: `Edit ${inputName}` }));
    const inputElement = screen.getByRole('textbox', { name: 'Edit name' });
    await user.clear(inputElement);
    await user.type(inputElement, 'New Name');
    await user.click(screen.getByRole('button', { name: `Save ${inputName}` }));
    expect(onSaveMock).toHaveBeenCalledWith('New Name');
  });

  it('shows an alert when trying to save an empty value', async () => {
    const user = userEvent.setup();
    const alertMock = vi.spyOn(window, 'alert').mockImplementation(() => {});
    render(<CustomInput placeholder={placeholder} inputName={inputName} value={initialValue} onSave={onSaveMock} />);
    await user.click(screen.getByRole('button', { name: `Edit ${inputName}` }));
    const inputElement = screen.getByRole('textbox', { name: 'Edit name' });
    await user.clear(inputElement);
    await user.click(screen.getByRole('button', { name: `Save ${inputName}` }));
    expect(alertMock).toHaveBeenCalledWith(`${inputName} cannot be empty!`);
    alertMock.mockRestore();
  });

  it('saves on blur', async () => {
    const user = userEvent.setup();
    render(<CustomInput placeholder={placeholder} inputName={inputName} value={initialValue} onSave={onSaveMock} />);

    await user.click(screen.getByRole('button', { name: `Edit ${inputName}` }));
    const inputElement = screen.getByRole('textbox', { name: 'Edit name' });
    await user.clear(inputElement);
    await user.type(inputElement, 'Another Name');
    fireEvent.blur(inputElement);
    expect(onSaveMock).toHaveBeenCalledWith('Another Name');
  });

  it('matches snapshot', () => {
    const { container } = render(
      <CustomInput placeholder={placeholder} inputName={inputName} value={initialValue} onSave={onSaveMock} />
    );
    expect(container).toMatchSnapshot();
  });
});
