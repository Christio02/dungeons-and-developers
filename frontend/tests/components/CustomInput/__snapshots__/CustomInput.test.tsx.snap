// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CustomInput Component > matches snapshot 1`] = `
<div>
  <section
    class="flex flex-row items-center justify-center gap-4 "
  >
    <h2
      aria-label="Name of your dungeon"
      class="header"
      data-testid="dungeon-name"
    >
      My dungeon
    </h2>
    <button
      aria-label="Edit Dungeon"
      class=" p-[2px] lg:p-[5px] rounded-lg bg-customRed hover:bg-transparent border-2 border-customRed hover:border-customRed transition-colors duration-100 group"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium text-white group-hover:text-customRed duration-100 lg:!w-6 lg:!h-6 xl:!w-8 xl:!h-8 css-20bmp1-MuiSvgIcon-root"
        data-testid="ModeEditOutlineIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 5.63l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83c.39-.39.39-1.02 0-1.41"
        />
      </svg>
    </button>
  </section>
</div>
`;
